### ISSUE TITLE
Issue when assigning pointers to structure's variable

### ISSUE THREAD

**Issue by sbacchio** [wlav/cppyy#8]
Hi Wim, again me :)

I've run into another issue with 2.1.0. 
It showed up first when assigning pointers-to-pointers (e.g. `int**`) to a structure's variable.

E.g. take the following:

```c++
struct Foo {
  int** bar;
};
```

```python
foo = gbl.Foo()
foo.bar = value
```

The latter would assign the wrong value incurring then into a seg fault.

When trying to achieve a reproducible, things became even weirder... (see [Explanation](#Explanation) after)

### Reproducible

```python
from cppyy import ll, cppdef, gbl

cppdef("""

template<typename T>
struct Foo {
  T* bar;

  Foo() {}

  Foo(T* other) : bar(other) { }

  bool eq(T* other) { return bar == other; }

  void print_check(char* msg, T* other) {
    printf("%s: %p, other: %p\\n", msg, bar, other);
  }
};

template<typename T>
auto create(T* other) {
  return Foo<T>(other);
}

""")

for dtype in ["int", "int*", "int**",]:

    print("\nType\n", dtype+"*")
    
    bar = ll.malloc[dtype](4)

    # Variable assignment
    foo = gbl.Foo[dtype]()
    foo.bar = bar
    foo.print_check("foo.bar", bar)

    # Pointer passed to the constructor
    foo2 = gbl.Foo[dtype](bar)
    foo2.print_check("foo2.bar", bar)
    
    # Pointer passed to a function
    foo3 = gbl.create[dtype](bar)
    foo3.print_check("foo3.bar", bar) 


    # These checks can be used for tests
    #assert foo.eq(bar)
    #assert foo2.eq(bar)
    #assert foo3.eq(bar)

    ll.free(bar)
```

### Explanation

I have tried three different approaches to initialize the structure:
- `foo`: as in the example above, first allocates the structure and then assigns via python
- `foo2`: assigns via a constructor
- `foo3`: assigns via a function

For a structure containing `int*` everything looks fine having an output:
```
Type int*

foo.bar: 0x55a1e2e30eb0, other: 0x55a1e2e30eb0
foo2.bar: 0x55a1e2e30eb0, other: 0x55a1e2e30eb0
foo3.bar: 0x55a1e2e30eb0, other: 0x55a1e2e30eb0
```
where all pointers, stored and original (`other`), are the same.

For a structure containing `int**` things look very odd, having all pointers different
```
Type int**

foo.bar: 0x55a1e2e85cb0, other: 0x7ffdaf0f9f30
foo2.bar: 0x7ffdaf0f9ea0, other: 0x7ffdaf0f9f30
foo3.bar: 0x7ffdaf0f9d50, other: 0x7ffdaf0f9f30
```

Finally, for a structure containing `int***`, I run into the following error...
```
Type int***

Traceback (most recent call last):
  File "tmp.py", line 35, in <module>
    foo.bar = bar
RuntimeError: property type mismatch or assignment not allowed
```

**Comment by sbacchio** [comment]
So yes, I cannot make so much sense of these outputs and I cannot say if some of them are expected and/or are known to not be working like for `int***`.
Would be great if you can have a look when you have some time!

**Comment by wlav** [comment]
The assignments and passing should work fine per se (I have a whole set of tests that do just that, up to 3D, see here: https://github.com/wlav/cppyy/blob/master/test/test_lowlevel.py#L506). The problem appears wholly with `bar = ll.malloc[dtype](4)`, as for the first two`bar`  thinks it's an integer array and for the last one a long array (hence the type mismatch; the long most obviously b/c it represents the pointer).

Must be something obvious missing as the generated malloc funcs aren't any different than the ones I use in the test ...

**Comment by wlav** [comment]
The problem is the reshaping. For the result to know that it is of length `4`, the `ll.malloc` call fixes up the size after the fact, by reshaping to `(4,)`. And that turns each multi-dim array into a 1D array. That's an easy fix, at least for the `int**` case, which seems to have the proper shape before the reshaping (is in repo; change is in ll.py, so easy to apply locally). The `int***` case, however, does not. Still digging for that one.

**Comment by wlav** [comment]
Okay, done. I had to refactor the executors same way as the converters to support any-dimensional pointer returns.

**Comment by sbacchio** [comment]
Wow that was super efficient! Thanks a lot!
So yes, in my case I wasn't using ll.malloc, but I was doing the same mistake with the reshaping.

**Comment by wlav** [comment]
Released with cppyy 2.2.0 and its dependencies.
